src/security/security-tests/src/test/java/org/geoserver/security/impl/DefaultDataAccessManagerTreeTest.java

```diff
@@ -135,4 +135,17 @@ public void testComplex() throws Exception {
         assertTrue(bases.canAccess(milUser, AccessMode.READ));
         assertTrue(bases.canAccess(milUser, AccessMode.WRITE));
     }
+
+    @Test
+    public void testBuildInFunctionResourceFilter() throws Exception {
+        DefaultResourceAccessManager defaultResourceAccessManager =
+                (DefaultResourceAccessManager) buildManager("complex.properties");
+        // validate that the filter is built correctly and there are no casting issues when
+        // populating layerExceptionIds
+        assertEquals(
+                "[[ NOT [ in([id], [arc.grid-lid]) = true ] ] AND [ NOT [ in([id], [bases-lid]) = true ] ]]",
+                defaultResourceAccessManager
+                        .buildInFunctionResourceFilter(rwUser, toppWs.getClass())
```

Comment:
toppWs is a proxy, it has a really weird class. Maybe better to have tests using explicit classes, like ResourceInfo.class, WorkspaceInfo.class

Reply:
fixed